internal/innerConcat.js

```diff
@@ -0,0 +1,26 @@
+/** @license ISC License (c) copyright 2017 original and current authors */
+/** @author Ian Hofmann-Hicks (evil) */
+
+const curry = require('../helpers/curry')
+const isSameType = require('../predicates/isSameType')
+const isSemigroup = require('../predicates/isSemigroup')
+
+function innerConcat(type, m) {
+  const t = type.type()
+
+  return function(left) {
+    if(!isSemigroup(left)) {
+      throw new TypeError(`${t}.concat: Both containers must contain Semigroups of the same type`)
+    }
+
+    return m.map(right => {
+      if(!isSameType(left, right)) {
+        throw new TypeError(`${t}.concat: Both containers must contain Semigroups of the same type`)
+      }
```

Comment:
If I think I understand what you are saying, I think you are saying that this would not throw an Error:
```js
const m = Maybe.of('hello')
const n = Maybe.of(3)

// will fail as Number is NOT a semigroup
m.concat(n)
```

if you look we are mapping on `m`, which is the container passed into us. So `right` will be a `Number` and it is not the same type as `String`, so this fails.

Now if you are saying that they are passing in the actual constructors `Array` and `String`
That would fail because `Array` is not a semigroup.

Is that what you mean?